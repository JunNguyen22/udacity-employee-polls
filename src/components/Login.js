import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import { useState } from "react";
import { useNavigate } from "react-router-dom";

const getErrorMessage = (id, password) => {
  if (!id || !password) {
    return "Please fill in User and Password.";
  }
  return "Incorrect User or Password, please try again.";
};

const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let navigate = useNavigate();
    return <Component {...props} router={{ navigate }} />;
  };

  return ComponentWithRouterProp;
};

const Login = ({ dispatch, router }) => {
  const [id, setId] = useState("");
  const [password, setPassword] = useState("");
  const [isSubmitted, setIsSubmitted] = useState(false);
  const [errorMessage, setErrorMessage] = useState("");

  const handleSubmit = () => {
    setIsSubmitted(true);
    setErrorMessage(getErrorMessage(id, password));
    if (id && password) {
      dispatch(setAuthedUser({ id, password }));
      router.navigate("/");
    }
  };

  return (
    <div className="login">
      <div className="card title-card">Login</div>
      <div className="card">User</div>
      <input placeholder="User" onChange={(e) => setId(e.target.value)} />
      <div className="card">Password</div>
      <input
        placeholder="Password"
        onChange={(e) => setPassword(e.target.value)}
        type="password"
      />
      <div className="card">
        <button
          data-testid="login-submit"
          className="btn"
          onClick={handleSubmit}
        >
          Login
        </button>
      </div>
      {isSubmitted ? (
        <div data-testid="login-error-message" className="error-text card">
          {errorMessage}
        </div>
      ) : null}
    </div>
  );
};

const mapStateToProps = () => ({});

export default withRouter(connect(mapStateToProps)(Login));

export const forTestings = {
  getErrorMessage,
};
