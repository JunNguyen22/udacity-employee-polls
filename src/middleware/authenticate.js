const authenticate = (store) => (next) => (action) => {
  let id;
  if (action.type === "SET_AUTHED_USER") {
    console.log("run authenticate: ", { ...action }, store.getState());
    const { users } = store.getState();
    const findUser = users[action?.id];
    console.log("finduser: ", findUser);
    if (findUser?.password === action?.password) {
      id = findUser?.id;
    }
  }

  //   console.group(action.type);
  //   console.log("Authenticate: ", action);
  //   const returnValue = next(action);
  //   console.log("The new state: ", store.getState());
  //   console.groupEnd();
  return next({ ...action, id });
};

export default authenticate;
