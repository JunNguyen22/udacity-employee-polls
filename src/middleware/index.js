import thunk from "redux-thunk";
import logger from "./logger";
import authenticate from "./authenticate";
import { applyMiddleware } from "redux";

export default applyMiddleware(thunk, logger, authenticate);
