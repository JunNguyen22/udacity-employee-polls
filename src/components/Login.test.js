import { render, fireEvent } from "@testing-library/react";
import Login from "./Login";
import { Provider } from "react-redux";
import reducer from "../reducers";
import middleware from "../middleware";
import { createStore } from "redux";
import "@testing-library/jest-dom";
import { forTestings } from "./Login";

const store = createStore(reducer, middleware);

describe("Login", () => {
  it("snapshot Login", () => {
    const component = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it("error message should show if User or Password isn't filled", async () => {
    const component = render(
      <Provider store={store}>
        <Login />
      </Provider>
    );

    const button = component.getByTestId("login-submit");
    fireEvent.click(button);
    const errorMessage = await component
      .findByTestId("login-error-message")
      .catch((val) => val);
    expect(errorMessage).toBeInTheDocument();
  });

  it("getErrorMessage has data", () => {
    const testValue = forTestings.getErrorMessage("test", "test");
    expect(testValue).toEqual("Incorrect User or Password, please try again.");
  });

  it("getErrorMessage doesn't have data", () => {
    const testValue = forTestings.getErrorMessage();
    expect(testValue).toEqual("Please fill in User and Password.");
  });
});
