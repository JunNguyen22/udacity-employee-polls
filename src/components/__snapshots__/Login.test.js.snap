// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot Login 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login"
      >
        <div
          class="card title-card"
        >
          Login
        </div>
        <div
          class="card"
        >
          User
        </div>
        <input
          placeholder="User"
        />
        <div
          class="card"
        >
          Password
        </div>
        <input
          placeholder="Password"
          type="password"
        />
        <div
          class="card"
        >
          <button
            class="btn"
            data-testid="login-submit"
          >
            Login
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login"
    >
      <div
        class="card title-card"
      >
        Login
      </div>
      <div
        class="card"
      >
        User
      </div>
      <input
        placeholder="User"
      />
      <div
        class="card"
      >
        Password
      </div>
      <input
        placeholder="Password"
        type="password"
      />
      <div
        class="card"
      >
        <button
          class="btn"
          data-testid="login-submit"
        >
          Login
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
